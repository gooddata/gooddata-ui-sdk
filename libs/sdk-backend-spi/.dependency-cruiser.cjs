const depCruiser = require("../../common/config/dep-cruiser/default.config");

options = {
    forbidden: [...depCruiser.DefaultRules, ...depCruiser.DefaultSdkRules],
    options: depCruiser.DefaultOptions,
};

module.exports = options;
