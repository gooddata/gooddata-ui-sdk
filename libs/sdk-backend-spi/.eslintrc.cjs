// (C) 2020 GoodData Corporation
module.exports = {
    plugins: ["sonarjs", "eslint-plugin-tsdoc", "regexp"],
    extends: [
        "@gooddata",
        "plugin:import-esm/recommended",
        "plugin:sonarjs/recommended",
        "plugin:regexp/recommended",
        "../../.eslintrc.js",
    ],
    overrides: [
        {
            files: ["*.ts", "*.tsx"],
            parser: "@typescript-eslint/parser",
            parserOptions: { tsconfigRootDir: __dirname, project: "tsconfig.json" },
        },
    ],
};
