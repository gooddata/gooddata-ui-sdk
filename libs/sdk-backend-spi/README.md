# Analytical Backend Service Provider Interface (SPI)

## License

(C) 2020-2021 GoodData Corporation

This project is under MIT License. See [LICENSE](LICENSE).
