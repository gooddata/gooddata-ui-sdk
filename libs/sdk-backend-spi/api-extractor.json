/**
 * Config file for API Extractor.  For more info, please visit: https://api-extractor.com
 */
{
    "$schema": "https://developer.microsoft.com/json-schemas/api-extractor/v7/api-extractor.schema.json",
    "extends": "../../common/config/api-extractor/api-extractor.json",
    "mainEntryPointFilePath": "<projectFolder>/esm/index.d.ts",
    "messages": {
        "extractorMessageReporting": {
            // we do not want to enforce this rule for now in this project, re-enable when dashboards are not alpha
            "ae-incompatible-release-tags": {
                "logLevel": "none",
                "addToApiReportFile": false
            }
        }
    }
}
