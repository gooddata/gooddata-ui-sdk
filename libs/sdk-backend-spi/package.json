{
    "name": "@gooddata/sdk-backend-spi",
    "version": "8.0.0-alpha.101",
    "author": "GoodData",
    "description": "GoodData Backend SPI abstraction interfaces",
    "repository": "https://github.com/gooddata/gooddata-ui-sdk/tree/master/libs/sdk-backend-spi",
    "license": "BSD-3-Clause",
    "main": "dist/index.js",
    "typings": "dist/index.d.ts",
    "files": [
        "dist",
        "src",
        "LICENSE"
    ],
    "config": {
        "tslint": "./{src,test}/**/*.ts"
    },
    "scripts": {
        "clean": "rm -rf ci dist coverage *.log && jest --clearCache",
        "dev": "tsc -p tsconfig.dev.json --watch",
        "build": "rm -rf dist && tsc -p tsconfig.build.json && npm run api-extractor",
        "api-extractor": "mkdir -p api && [ -z \"${CI}\" ] && (api-extractor run -l) || (api-extractor run)",
        "test": "jest --watch",
        "test-once": "jest",
        "test-ci": "JEST_JUNIT_OUTPUT=./ci/results/test-results.xml jest --ci --config jest.ci.js",
        "tslint": "tslint -p . -t verbose $npm_package_config_tslint",
        "tslint-ci": "mkdir -p ./ci/results && tslint -p . -t checkstyle -o ./ci/results/tslint-results.xml $npm_package_config_tslint",
        "prettier-check": "prettier --check '{src,test}/**/*.{ts,tsx,json,scss,md,yaml,html}'",
        "prettier-write": "prettier --write '{src,test}/**/*.{ts,tsx,json,scss,md,yaml,html}'",
        "dep-cruiser": "depcruise --validate .dependency-cruiser.js --output-type err-long src/",
        "dep-cruiser-ci": "depcruise --validate .dependency-cruiser.js --output-type err-long src/",
        "validate": "npm run dep-cruiser && npm run tslint && npm run prettier-check",
        "validate-ci": "npm run dep-cruiser-ci && npm run tslint-ci && npm run prettier-check"
    },
    "dependencies": {
        "@gooddata/sdk-model": "^8.0.0-alpha.101",
        "lodash": "^4.17.15",
        "spark-md5": "^3.0.0",
        "ts-invariant": "^0.4.4",
        "tslib": "^2.0.0"
    },
    "devDependencies": {
        "@gooddata/tslint-config": "1.0.0",
        "@microsoft/api-extractor": "^7.3.8",
        "@types/jest": "^25.1.2",
        "@types/lodash": "^4.14.123",
        "@types/spark-md5": "^3.0.1",
        "dependency-cruiser": "^8.0.1",
        "jest": "^25.1.0",
        "jest-junit": "^3.0.0",
        "prettier": "~2.0.5",
        "ts-jest": "^25.2.0",
        "tslint": "^5.14.0",
        "tslint-config-prettier": "^1.18.0",
        "typescript": "3.9.5"
    }
}
