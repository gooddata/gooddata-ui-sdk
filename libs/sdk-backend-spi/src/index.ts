// (C) 2019-2024 GoodData Corporation
/**
 * This package provides definitions of the Service Provider Interface (SPI) for the Analytical Backend.
 *
 * @remarks
 * The interface defines functionality to be implemented for a particular backend to be used in GoodData.UI.
 * The Analytical Backend SPI for GoodData Cloud and GoodData.CN (joint codename `tiger` in `@gooddata/sdk-backend-tiger`) is almost fully implemented.
 *
 * @packageDocumentation
 */
export {
    IAnalyticalBackend,
    IAnalyticalBackendConfig,
    AnalyticalBackendFactory,
    prepareExecution,
    IAuthenticationContext,
    IAuthenticatedPrincipal,
    IAuthenticationProvider,
    NotAuthenticatedHandler,
} from "./backend/index.js";

export { IBackendCapabilities } from "./backend/capabilities.js";

export { IUserSettings, IWorkspaceSettings, IUserWorkspaceSettings } from "./common/settings.js";

export { IUserService } from "./user/index.js";
export { IUserSettingsService } from "./user/settings/index.js";

export {
    IExecutionFactory,
    IPreparedExecution,
    IExecutionResult,
    IDataView,
    ExplainConfig,
    IExplainResult,
    IExplainProvider,
    ExplainType,
} from "./workspace/execution/index.js";

export { IWorkspaceSettingsService } from "./workspace/settings/index.js";

export { CancelableOptions, ICancelable } from "./cancelation/index.js";

export {
    IGetInsightOptions,
    IGetVisualizationClassesOptions,
    IWorkspaceInsightsService,
    InsightOrdering,
    IInsightsQueryOptions,
    IInsightsQuery,
    IInsightsQueryResult,
    IInsightReferences,
    IInsightReferencing,
    InsightReferenceTypes,
    SupportedInsightReferenceTypes,
} from "./workspace/insights/index.js";

export {
    ExportDefinitionOrdering,
    IExportDefinitionsQuery,
    IExportDefinitionsQueryOptions,
    IExportDefinitionsQueryResult,
    IGetExportDefinitionOptions,
    IWorkspaceExportDefinitionsService,
    ExportDefinitionQuerySortDirection,
    ExportDefinitionQuerySortProperty,
    ExportDefinitionQuerySort,
} from "./workspace/exportDefinitions/index.js";

export {
    IElementsQueryFactory,
    IElementsQueryResult,
    IElementsQuery,
    IElementsQueryOptions,
    IElementsQueryAttributeFilter,
    IElementsQueryOptionsElementsByUri,
    IElementsQueryOptionsElementsByValue,
    IElementsQueryOptionsElementsByPrimaryDisplayFormValue,
    ElementsQueryOptionsElementsSpecification,
    isElementsQueryOptionsElementsByValue,
    isElementsQueryOptionsElementsByPrimaryDisplayFormValue,
    isValueBasedElementsQueryOptionsElements,
    IFilterElementsQuery,
    FilterWithResolvableElements,
} from "./workspace/attributes/elements/index.js";

export { IExportConfig, IExportResult } from "./workspace/execution/export.js";

export { IWorkspaceStylingService } from "./workspace/styling/index.js";
export {
    AnalyticalBackendError,
    NoDataError,
    DataTooLargeError,
    TimeoutError,
    ProtectedDataError,
    UnexpectedResponseError,
    UnexpectedError,
    NotSupported,
    NotImplemented,
    NotAuthenticated,
    NotAuthenticatedReason,
    LimitReached,
    ContractExpired,
    ErrorConverter,
    isAnalyticalBackendError,
    isNoDataError,
    isDataTooLargeError,
    isProtectedDataError,
    isUnexpectedResponseError,
    isUnexpectedError,
    isNotSupported,
    isNotImplemented,
    isNotAuthenticated,
    isLimitReached,
    isContractExpired,
    AuthenticationFlow,
    AnalyticalBackendErrorTypes,
} from "./errors/index.js";

export { IPagedResource } from "./common/paging.js";

export {
    IAnalyticalWorkspace,
    IWorkspacesQuery,
    IWorkspacesQueryFactory,
    IWorkspacesQueryResult,
    IWorkspaceDescriptor,
} from "./workspace/index.js";

export { IAttributeHierarchiesService } from "./workspace/attributeHierarchies/index.js";

export { IWorkspacePermissionsService } from "./workspace/permissions/index.js";

export { IWorkspaceAttributesService } from "./workspace/attributes/index.js";

export { IWorkspaceMeasuresService, IMeasureReferencing } from "./workspace/measures/index.js";

export { IWorkspaceFactsService } from "./workspace/facts/index.js";

export {
    IWorkspaceDashboardsService,
    IGetDashboardOptions,
    IGetDashboardPluginOptions,
    IGetScheduledMailOptions,
    IWidgetAlertCount,
    SupportedDashboardReferenceTypes,
    IWidgetReferences,
    SupportedWidgetReferenceTypes,
    IDashboardReferences,
    IDashboardWithReferences,
    IDashboardsQuery,
    IDashboardsQueryResult,
} from "./workspace/dashboards/index.js";
export {
    isDashboardLayoutEmpty,
    IWidgetWithLayoutPath,
    LayoutPath,
    layoutWidgets,
    layoutWidgetsWithPaths,
    walkLayout,
} from "./workspace/dashboards/utils.js";
export {
    IWorkspaceUsersQuery,
    IWorkspaceUsersQueryOptions,
    IWorkspaceUsersQueryResult,
} from "./workspace/users/index.js";
export {
    IDateFilterConfigsQuery,
    IDateFilterConfigsQueryResult,
} from "./workspace/dateFilterConfigs/index.js";

export {
    IWorkspaceCatalogFactory,
    IWorkspaceCatalogAvailableItemsFactory,
    IWorkspaceCatalog,
    IWorkspaceCatalogFactoryOptions,
    IWorkspaceCatalogWithAvailableItems,
    IWorkspaceCatalogWithAvailableItemsFactoryOptions,
    IWorkspaceCatalogFactoryMethods,
    IWorkspaceCatalogMethods,
} from "./workspace/ldm/catalog.js";

export { IWorkspaceDatasetsService } from "./workspace/ldm/datasets.js";

export {
    IMeasureExpressionToken,
    IObjectExpressionToken,
    IAttributeElementExpressionToken,
    ITextExpressionToken,
    ICommentExpressionToken,
    IBracketExpressionToken,
} from "./workspace/measures/measure.js";

export { IOrganization, IOrganizations } from "./organization/index.js";
export { IEntitlements } from "./entitlements/index.js";
export { IDataSourcesService } from "./dataSources/index.js";
export { ISecuritySettingsService, ValidationContext } from "./organization/securitySettings/index.js";
export { IOrganizationStylingService } from "./organization/styling/index.js";
export { IOrganizationSettingsService } from "./organization/settings/index.js";

export {
    IWorkspaceUserGroupsQuery,
    IWorkspaceUserGroupsQueryOptions,
    IWorkspaceUserGroupsQueryResult,
} from "./workspace/userGroups/index.js";

export { IWorkspaceAccessControlService } from "./workspace/accessControl/index.js";

export {
    IOrganizationUserService,
    IOrganizationUsersQuery,
    IOrganizationUsersQueryResult,
    IOrganizationUserGroupsQuery,
    IOrganizationUserGroupsQueryResult,
} from "./organization/users/index.js";
export { IOrganizationPermissionService, IPermissionsAssignment } from "./organization/permissions/index.js";
