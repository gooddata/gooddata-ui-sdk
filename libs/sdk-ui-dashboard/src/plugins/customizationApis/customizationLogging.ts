// (C) 2021-2022 GoodData Corporation

import { IDashboardPluginContract_V1 } from "../plugin.js";
import { pluginDebugStr } from "./pluginUtils.js";

function addPluginInfoToMessage(plugin: IDashboardPluginContract_V1 | undefined, message: string) {
    return plugin ? `${pluginDebugStr(plugin)}: ${message}` : message;
}

export interface IDashboardCustomizationLogger {
    setCurrentPlugin(plugin: IDashboardPluginContract_V1 | undefined): void;

    log(message: string, ...optionalParams: any[]): void;
    warn(message: string, ...optionalParams: any[]): void;
    error(message: string, ...optionalParams: any[]): void;
}

/**
 * Common logger to use for all events that occur during customization. The logger is responsible for adding
 * information about plugin whose registration code triggered those events.
 */
export class DashboardCustomizationLogger implements IDashboardCustomizationLogger {
    private currentPlugin: IDashboardPluginContract_V1 | undefined;

    public setCurrentPlugin = (plugin: IDashboardPluginContract_V1 | undefined): void => {
        this.currentPlugin = plugin;
    };

    public log = (message: string, ...optionalParams: any[]): void => {
        // eslint-disable-next-line no-console
        console.log(addPluginInfoToMessage(this.currentPlugin, message), optionalParams);
    };
    public warn = (message: string, ...optionalParams: any[]): void => {
        console.warn(addPluginInfoToMessage(this.currentPlugin, message), optionalParams);
    };
    public error = (message: string, ...optionalParams: any[]): void => {
        console.error(addPluginInfoToMessage(this.currentPlugin, message), optionalParams);
    };
}
