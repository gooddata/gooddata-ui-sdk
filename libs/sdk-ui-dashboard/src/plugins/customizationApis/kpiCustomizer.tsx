// (C) 2021 GoodData Corporation
import { IDashboardKpiCustomizer } from "../customizer";
import { DefaultDashboardKpi, KpiComponentProvider, OptionalKpiComponentProvider } from "../../presentation";
import { InvariantError } from "ts-invariant";
import { DashboardCustomizationLogger } from "./customizationLogging";

const DefaultKpiRendererProvider: KpiComponentProvider = () => {
    return DefaultDashboardKpi;
};

interface IKpiCustomizerState {
    addCustomProvider(provider: OptionalKpiComponentProvider): void;
    getRootProvider(): KpiComponentProvider;
    switchRootProvider(provider: KpiComponentProvider): void;
}

class DefaultKpiCustomizerState implements IKpiCustomizerState {
    /*
     * Chain of 'core' providers. Core providers are evaluated from last to first. As soon as some provider
     * returns a Component, then that component will be used for rendering. If provider returns undefined,
     * the evaluation continues to next provider in the chain.
     *
     * Note: the chain is 'primed' with a provider that always returns the default KPI renderer. This is
     * essential to allow the decorations to work - purely because decorations can only work if there is
     * something to decorate. See constructor.
     */
    private readonly coreProviderChain: KpiComponentProvider[];

    /*
     * Core provider encapsulates resolution using the chain of core providers.
     */
    private readonly coreProvider: KpiComponentProvider = (kpi, widget) => {
        const providerStack = [...this.coreProviderChain].reverse();

        for (const provider of providerStack) {
            const Component = provider(kpi, widget);

            if (Component) {
                return Component;
            }
        }

        // if this happens then the provider chain got messed up. by default the chain contains the default
        // provider which never returns undefined
        throw new InvariantError();
    };

    /*
     * Root provider is THE provider that should be used in the dashboard extension properties. The
     * provider function included here will reflect the setup where there may be N registered decorators
     * sitting on top of a chain of core providers.
     *
     * In the initial state the root provider IS the core provider - meaning no decorations. As the
     * decorators get registered, the rootProvider changes.
     */
    private rootProvider: KpiComponentProvider = this.coreProvider;

    constructor(defaultProvider: KpiComponentProvider) {
        this.coreProviderChain = [defaultProvider];
    }

    addCustomProvider(provider: KpiComponentProvider): void {
        this.coreProviderChain.push(provider);
    }

    getRootProvider(): KpiComponentProvider {
        return this.rootProvider;
    }

    switchRootProvider(provider: KpiComponentProvider): void {
        this.rootProvider = provider;
    }
}

/**
 * Sealed customizer state will not allow to perform any modifications of the state. Sealing the state
 * is essential to make the customization more defensive: during the registration, the plugin can hang
 * onto the dashboard customizer - stash it somewhere. And then after registration try to use the customizer
 * and try to do additional 'ad-hoc' customizations.
 *
 * Such a thing is not supported by the lifecycle and this sealing is in place to prevent plugins going into
 * that dangerous territory.
 */
class SealedKpiCustomizerState implements IKpiCustomizerState {
    constructor(
        private readonly logger: DashboardCustomizationLogger,
        private readonly state: IKpiCustomizerState,
    ) {}

    public addCustomProvider = (_provider: KpiComponentProvider): void => {
        // eslint-disable-next-line no-console
        this.logger.warn(`Attempting to customize KPI rendering outside of plugin registration. Ignoring.`);
    };

    // eslint-disable-next-line sonarjs/no-identical-functions
    public switchRootProvider = (_provider: KpiComponentProvider): void => {
        // eslint-disable-next-line no-console
        this.logger.warn(`Attempting to customize KPI rendering outside of plugin registration. Ignoring.`);
    };

    public getRootProvider = (): KpiComponentProvider => {
        return this.state.getRootProvider();
    };
}

/**
 * Default implementation of the KPI rendering customizer. Notice that the state of the customizations
 * is kept separate from this class.
 *
 * This code is responsible for creation of the providers (if needed) and then updating the state
 * accordingly. The customizer state methods are responsible for doing correct updates of the state itself. This
 * decoupling is in place so that it is possible to seal the state and prevent write operations from some point
 * onward. See {@link SealedKpiCustomizerState} for more motivation behind this.
 *
 * @internal
 */
export class DefaultKpiCustomizer implements IDashboardKpiCustomizer {
    private readonly logger: DashboardCustomizationLogger;
    private state: IKpiCustomizerState;

    constructor(
        logger: DashboardCustomizationLogger,
        defaultProvider: KpiComponentProvider = DefaultKpiRendererProvider,
    ) {
        this.logger = logger;
        this.state = new DefaultKpiCustomizerState(defaultProvider);
    }

    public withCustomProvider = (provider: OptionalKpiComponentProvider): IDashboardKpiCustomizer => {
        this.state.addCustomProvider(provider);

        return this;
    };

    public withCustomDecorator = (
        providerFactory: (next: KpiComponentProvider) => OptionalKpiComponentProvider,
    ): IDashboardKpiCustomizer => {
        // snapshot current root provider
        const rootSnapshot = this.state.getRootProvider();
        // call user's factory in order to obtain the actual provider - pass the current root so that user's
        // code can use it to obtain component to decorate
        const decoratorProvider = providerFactory(rootSnapshot);
        // construct new root provider; this will be using user's provider with a fallback to root provider
        // in case user's code does not return anything
        const newRootProvider: KpiComponentProvider = (kpi, widget) => {
            const Component = decoratorProvider(kpi, widget);

            if (Component) {
                return Component;
            }

            return rootSnapshot(kpi, widget);
        };

        // finally modify the root provider; next time someone registers decorator, it will be on top of
        // this currently registered one
        this.state.switchRootProvider(newRootProvider);

        return this;
    };

    public getKpiProvider = (): KpiComponentProvider => {
        return this.state.getRootProvider();
    };

    public sealCustomizer = (): void => {
        this.state = new SealedKpiCustomizerState(this.logger, this.state);
    };
}
