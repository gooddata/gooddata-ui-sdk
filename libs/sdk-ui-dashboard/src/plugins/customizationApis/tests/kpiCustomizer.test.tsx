// (C) 2021 GoodData Corporation

import React from "react";
import {
    IDashboardKpiProps,
    KpiComponentProvider,
    OptionalKpiComponentProvider,
} from "../../../presentation";
import { idRef, measureItem } from "@gooddata/sdk-model";
import { IKpiWidget, ILegacyKpi } from "@gooddata/sdk-backend-spi";
import { ReferenceLdm } from "@gooddata/reference-workspace";
import { mount } from "enzyme";
import invariant from "ts-invariant";
import { DefaultKpiCustomizer } from "../kpiCustomizer";
import { DashboardCustomizationLogger } from "../customizationLogging";

//
//
//

const TestKpiWidget: IKpiWidget = {
    type: "kpi",
    title: "Test KPI",
    ref: idRef("testKpi"),
    identifier: "testKpi",
    description: "Test KPI Widget",
    uri: "/uri/testKpi",
    kpi: {
        comparisonDirection: "growIsGood",
        metric: measureItem(ReferenceLdm.Amount),
        comparisonType: "previousPeriod",
    },
    drills: [],
    ignoreDashboardFilters: [],
};
const CustomTitle = "TestTitle";
const TestKpiWidgetWithCustomTitle: IKpiWidget = { ...TestKpiWidget, title: CustomTitle };

//
//
//

function createTestComponent(name: string): React.FC {
    function TestComponent() {
        return <div id={name} />;
    }

    return TestComponent;
}

function createTestComponentProvider(
    name: string,
    predicate: (kpi: ILegacyKpi, widget: IKpiWidget) => boolean,
): OptionalKpiComponentProvider {
    const Component = createTestComponent(name);

    return (kpi, widget) => {
        if (!predicate(kpi, widget)) {
            return undefined;
        }

        return Component;
    };
}

function createTestDecoratorFactory(
    name: string,
    predicate: (kpi: ILegacyKpi, widget: IKpiWidget) => boolean,
) {
    return (next: KpiComponentProvider): OptionalKpiComponentProvider => {
        return (kpi, widget) => {
            if (!predicate(kpi, widget)) {
                return undefined;
            }

            function Decorator(props: IDashboardKpiProps) {
                const Decorated = next(kpi, widget)!;

                return (
                    <div id={name}>
                        <Decorated {...props} />
                    </div>
                );
            }

            return Decorator;
        };
    };
}

const DefaultTestComponentProvider: KpiComponentProvider = createTestComponentProvider(
    "default",
    () => true,
) as KpiComponentProvider;

//
//
//

function renderToHtml(customizer: DefaultKpiCustomizer, widget: IKpiWidget) {
    const provider = customizer.getKpiProvider();
    const Component = provider(widget.kpi, widget);
    const dummyProps: IDashboardKpiProps = { dummyKpiProps: true } as any;

    // this should not happen; if it does something is seriously hosed in the customizer
    invariant(Component);

    const wrapper = mount(<Component {...dummyProps} />);
    return wrapper.html();
}

describe("KPI customizer", () => {
    let Customizer: DefaultKpiCustomizer;

    beforeEach(() => {
        Customizer = new DefaultKpiCustomizer(
            new DashboardCustomizationLogger(),
            DefaultTestComponentProvider,
        );
    });

    it("should fallback to rendering default component", () => {
        expect(renderToHtml(Customizer, TestKpiWidget)).toMatchSnapshot();
    });

    describe("withCustomProvider", () => {
        it("should work for KPI that matches predicate", () => {
            const provider = createTestComponentProvider(
                "fromCustomProvider",
                (_kpi, widget) => widget.title === CustomTitle,
            );
            Customizer.withCustomProvider(provider);

            expect(renderToHtml(Customizer, TestKpiWidgetWithCustomTitle)).toMatchSnapshot();
        });

        it("should fallback to default component if insight does not match provider criteria", () => {
            const provider = createTestComponentProvider(
                "fromCustomProvider",
                (_kpi, widget) => widget.title === CustomTitle,
            );
            Customizer.withCustomProvider(provider);

            expect(renderToHtml(Customizer, TestKpiWidget)).toMatchSnapshot();
        });

        it("should use component from latest registered provider that matches the insight", () => {
            const provider1 = createTestComponentProvider(
                "fromCustomProvider1",
                (_kpi, widget) => widget.title === CustomTitle,
            );
            const provider2 = createTestComponentProvider(
                "fromCustomProvider2",
                (_kpi, widget) => widget.title === CustomTitle,
            );
            const provider3 = createTestComponentProvider(
                "fromCustomProvider3",
                (_kpi, widget) => widget.title === CustomTitle,
            );
            Customizer.withCustomProvider(provider1);
            Customizer.withCustomProvider(provider2);
            Customizer.withCustomProvider(provider3);

            // component from last registered and matching provider is `fromCustomProvider3`
            expect(renderToHtml(Customizer, TestKpiWidgetWithCustomTitle)).toMatchSnapshot();
        });
    });

    describe("withCustomDecorator", () => {
        it("should decorate default component if no other providers are registered", () => {
            const factory = createTestDecoratorFactory("decorator1", () => true);
            Customizer.withCustomDecorator(factory);

            expect(renderToHtml(Customizer, TestKpiWidget)).toMatchSnapshot();
        });

        it("should decorate custom component registered for an insight", () => {
            const factory = createTestDecoratorFactory("decorator1", () => true);
            Customizer.withCustomDecorator(factory);

            const provider = createTestComponentProvider(
                "fromCustomProvider",
                (_kpi, widget) => widget.title === CustomTitle,
            );
            Customizer.withCustomProvider(provider);

            expect(renderToHtml(Customizer, TestKpiWidgetWithCustomTitle)).toMatchSnapshot();
        });

        it("should decorate default component if insight does not match custom component criteria ", () => {
            const factory = createTestDecoratorFactory("decorator1", () => true);
            Customizer.withCustomDecorator(factory);
            const provider = createTestComponentProvider(
                "fromCustomProvider",
                (_kpi, widget) => widget.title === CustomTitle,
            );
            Customizer.withCustomProvider(provider);

            expect(renderToHtml(Customizer, TestKpiWidget)).toMatchSnapshot();
        });

        it("should decorate if insight matches criteria", () => {
            const factory = createTestDecoratorFactory(
                "decorator1",
                (_kpi, widget) => widget.title === CustomTitle,
            );
            Customizer.withCustomDecorator(factory);

            expect(renderToHtml(Customizer, TestKpiWidgetWithCustomTitle)).toMatchSnapshot();
        });

        it("should not decorate if insight does not match criteria", () => {
            const factory = createTestDecoratorFactory(
                "decorator1",
                (_kpi, widget) => widget.title === CustomTitle,
            );
            Customizer.withCustomDecorator(factory);

            expect(renderToHtml(Customizer, TestKpiWidget)).toMatchSnapshot();
        });

        it("should use multiple decorators if insight matches criteria", () => {
            const factory1 = createTestDecoratorFactory(
                "decorator1",
                (_kpi, widget) => widget.title === CustomTitle,
            );
            const factory2 = createTestDecoratorFactory(
                "decorator2",
                (_kpi, widget) => widget.title === CustomTitle,
            );
            Customizer.withCustomDecorator(factory1);
            Customizer.withCustomDecorator(factory2);

            // decoration starts from last register; so need to see decorator2 -> decorator1 -> default component
            expect(renderToHtml(Customizer, TestKpiWidgetWithCustomTitle)).toMatchSnapshot();
        });

        it("should use multiple decorators and custom component if insight matches criteria", () => {
            const factory1 = createTestDecoratorFactory(
                "decorator1",
                (_kpi, widget) => widget.title === CustomTitle,
            );
            const factory2 = createTestDecoratorFactory(
                "decorator2",
                (_kpi, widget) => widget.title === CustomTitle,
            );
            const provider = createTestComponentProvider(
                "customProvider1",
                (_kpi, widget) => widget.title === CustomTitle,
            );
            Customizer.withCustomDecorator(factory1);
            Customizer.withCustomDecorator(factory2);
            Customizer.withCustomProvider(provider);

            // decoration starts from last register; so need to see decorator2 -> decorator1 -> custom component
            expect(renderToHtml(Customizer, TestKpiWidgetWithCustomTitle)).toMatchSnapshot();
        });

        it("should only use only those decorators that match criteria and then use custom component", () => {
            const factory1 = createTestDecoratorFactory("decorator1", () => false);
            const factory2 = createTestDecoratorFactory(
                "decorator2",
                (_kpi, widget) => widget.title === CustomTitle,
            );
            const provider = createTestComponentProvider(
                "customProvider1",
                (_kpi, widget) => widget.title === CustomTitle,
            );

            Customizer.withCustomDecorator(factory1);
            Customizer.withCustomDecorator(factory2);
            Customizer.withCustomProvider(provider);

            // This will use decorator2 on top of customProvider1
            expect(renderToHtml(Customizer, TestKpiWidgetWithCustomTitle)).toMatchSnapshot();
        });
    });
});
