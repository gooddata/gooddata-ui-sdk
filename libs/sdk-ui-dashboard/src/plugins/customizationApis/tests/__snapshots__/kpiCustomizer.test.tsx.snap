// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`KPI customizer > should fallback to rendering default component 1`] = `"<div id="default"></div>"`;

exports[`KPI customizer > withCustomDecorator > should decorate custom component registered for an insight 1`] = `"<div id="decorator1"><div id="fromCustomProvider"></div></div>"`;

exports[`KPI customizer > withCustomDecorator > should decorate default component if insight does not match custom component criteria  1`] = `"<div id="decorator1"><div id="default"></div></div>"`;

exports[`KPI customizer > withCustomDecorator > should decorate default component if no other providers are registered 1`] = `"<div id="decorator1"><div id="default"></div></div>"`;

exports[`KPI customizer > withCustomDecorator > should decorate if insight matches criteria 1`] = `"<div id="decorator1"><div id="default"></div></div>"`;

exports[`KPI customizer > withCustomDecorator > should not decorate if insight does not match criteria 1`] = `"<div id="default"></div>"`;

exports[`KPI customizer > withCustomDecorator > should only use only those decorators that match criteria and then use custom component 1`] = `"<div id="decorator2"><div id="customProvider1"></div></div>"`;

exports[`KPI customizer > withCustomDecorator > should use multiple decorators and custom component if insight matches criteria 1`] = `"<div id="decorator2"><div id="decorator1"><div id="customProvider1"></div></div></div>"`;

exports[`KPI customizer > withCustomDecorator > should use multiple decorators if insight matches criteria 1`] = `"<div id="decorator2"><div id="decorator1"><div id="default"></div></div></div>"`;

exports[`KPI customizer > withCustomProvider > should fallback to default component if insight does not match provider criteria 1`] = `"<div id="default"></div>"`;

exports[`KPI customizer > withCustomProvider > should use component from latest registered provider that matches the insight 1`] = `"<div id="fromCustomProvider3"></div>"`;

exports[`KPI customizer > withCustomProvider > should work for KPI that matches predicate 1`] = `"<div id="fromCustomProvider"></div>"`;
